docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,153 @@
+
+However today, users are forced to run a sequence of commands tailored to only work with the proper directory structure on the system.
+
+## Major Goals
+
+1. Modularize the InstructLab workflow such that any part can be run independently
+2. Allow users to choose whether or not to take advantage of the config/system-profile method of running InstructLab. Meaning they do not need any pre-existing configuration to run the SDK.
+3. Standardize user contracts for the existing functionality of the InstructLab workflow. Existing CLI commands should be using the SDK once past click parsing, not separate code.
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
+
+1. Exposing all library functionality immediately
+2. Replacing CLI
+3. Shipping an SDK that is generally available as opposed to v1alpha1 or v1beta1.
+
+## Design
+
+### Versioning
+
+The SDK would start at version v1alpha1 such that it can change/break at any time for the first few iterations as libraries adjust their API surface.
+
+### Structure
+
+This SDK should live in a net new package inside of `instructlab/instructlab` preferably to limit unnecessary imports in a new repository. The SDK could be imported as `instructlab.core...`
+
+The user surface initially should look like this:
+
+`instructlab.core` contains all SDK definitions. Users can `from instructlab.core import...` to use specific SDK classes
+
+For most of the existing InstructLab command groups, there should be a class:
+
+`from instructlab.core import, Config, Taxonomy, Eval, Data, Model`
+
+The full list of classes and their methods for now:
+
+```console
+instructlab.core.Config
+instructlab.core.Config.init
+instructlab.core.Config.show (get)
+instructlab.core.Taxonomy
+instructlab.core.Taxonomy.diff
+instructlab.core.System
+instructlab.core.System.info
+instructlab.core.Data
+instructlab.core.Data.ingest
+instructlab.core.Data.generate_data
+instructlab.core.Model
+instructlab.core.Model.serve
+instructlab.core.Model.train_model
+instructlab.core.Model.process_data (calling the training library's data process class in a safe way)
+instructlab.core.Model.evaluate_mt_bench
+instructlab.core.Model.evaluate_dk_bench
+instructlab.core.Model.evaluate_mmlu_bench
+instructlab.core.RAG.ingest
+instructlab.core.RAG.convert
+instructlab.core.Process
+instructlab.core.Process.list
+instructlab.core.Process.attach
+```
+
+a brief example:
+
+```python
+
+from instructlab.core import Config, Taxonomy, Data, Model
+
+config_object = Config.init(...)
+diff = Taxonomy.diff()
+
+if diff:
+    # not in v1alpha1
+    data_client = Data(data_path="",..., teacher_model="", num_cpus="", taxonomy_path="",)
+    data_path = data_client.ingest()
+    # not in v1alpha1
+
+    openai_compat_client = some_server()
+
+    data_jsonls = data_client.generate_data(client=openai_compat_client, data=data_path)
```

Comment:
I wonder how you distinguish between parameters passed into verb methods (`client`, `data` here) and parameters passed to `class.__init__` (line 100).

Reply:
I would probably do something like:

```python
def generate_data(self, data=None...):
data = data if data is not None else self.data
...

```

something like "hey, use the class defaults if the value we are given is None. Also, I think we can hash things like this out during development as its hard to figure out these patterns until writing the classes I find